packages/modal/src/Modal.tsx

```diff
@@ -0,0 +1,281 @@
+   *
+   */
+  onRequestClose?: () => void;
+
+  /**
+   * className applied to root overlay div.
+   *
+   */
+  className?: string;
+}
+
+/**
+ * # Modal
+ *
+ *  Modals place content on top of main window.
+ *
+```
+<Modal
+  active
+  usePortal
+  size="large"
+  setActive={setActive}
+  title="My Modal"
+  onRequestClose={() => console.log('Modal is closing now!')}
+  >  
+  Modal content!
+</Modal>
+```
+ * @param props.active Boolean to describe whether or not Modal is active.
+ * @param props.usePortal Boolean to describe if content should be portaled to end of DOM, or appear in React DOM tree.
+ * @param props.size String to determine size of Modal. ['xxsmall', 'xsmall', 'small', 'normal', 'large', 'xlarge']
+ * @param props.setActive Callback to change the active state of Modal.
+ * @param props.children Content to appear inside of Modal container.
+ * @param props.title Title for the Modal, will appear inside header tags.
+ * @param props.onRequestClose Callback invoked when Modal is closed.
+ * @param props.className className applied to overlay div.
+ *
+ */
+function Modal({
+  active = false,
+  usePortal = true,
+  size = ModalSize.Normal,
+  setActive,
+  children,
+  title,
+  onRequestClose,
+  className,
+  ...rest
+}: ModalProps) {
+  const contentRef = useRef<HTMLDivElement>(null);
+
+  const handleClose = () => {
+    if (onRequestClose) {
+      onRequestClose();
+    }
+
+    if (setActive) {
+      setActive(false);
+    }
+  };
+
+  const handleEscape = useCallback((e: KeyboardEvent) => {
+    if (e.keyCode === EscapeKey) {
+      handleClose();
+    }
+  }, []);
+
+  const handleDocumentClick = (e: React.SyntheticEvent) => {
+    if (!contentRef.current) {
+      return;
+    }
+
+    if (!contentRef.current.contains(e.target as Node)) {
+      handleClose();
+    }
+  };
+
+  useEventListener('keydown', handleEscape, {
+    options: { once: true },
+  });
+
+  const Root = usePortal ? Portal : Fragment;
+
+  return active ? (
+    <Root>
+      <div {...rest} className={overlayStyle} onClick={handleDocumentClick}>
+        <div
+          className={cx(modalContentStyle, modalSizes[size], className)}
+          tabIndex={-1}
+          ref={contentRef}
+        >
+          <Icon
+            glyph="X"
+            fill={'#000'}
+            size={Size.Large}
+            onClick={handleClose}
+            className={closeButton}
+            data-dismiss="modal"
+            aria-hidden="true"
+          />
+
+          {title && (
+            <header>
+              <h3 className={titleStyle}>{title}</h3>
+              <hr />
+            </header>
+          )}
+
+          <div className={modalBodyStyle}>{children}</div>
+        </div>
+      </div>
+    </Root>
+  ) : null;
+}
+
+Modal.displayName = 'Modal';
+
+Modal.propTypes = {
+  active: PropTypes.bool,
+  usePortal: PropTypes.bool,
+  size: PropTypes.string,
+  children: PropTypes.node,
+  onRequestClose: PropTypes.func,
```

Comment:
What if this was `onWillClose`, and if the consumer returns false, it prevents the modal from closing?